en/lessons/basics/types.md

```diff
@@ -0,0 +1,177 @@
+---
+version: 1.0.0
+title: Types
+---
+
+Types are a fundamental and integral part of Haskell. Haskell's advanced 
+type system sets it apart from other languages, and is a powerful method for
+ensuring that your code does what you want it to do.
+
+{% include toc.html %}
+
+## Extensions and Imports
+
+You will need to add the following lines to the top of your Haskell source 
+file (`.hs`) to be able to run these examples:
+
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+import Data.Text ( Text )
+```
+
+## Static Typing
+
+Haskell is a statically-typed programming language. This means that the type
+of all data must be known before the program is compiled. Some imperative 
+languages, such as Java and C++ are also statically-typed, while others like 
+Python and JavaScript are not.
+
+## Type Inference
+
+Haskell's type system works very differently from that of Java and C++. 
+Instead of declaring the type of every single function and variable, Haskell
+can *infer* the type of values based on just a few declarations.
+
+Typically, you just need to specify the type of a function. The types of all
+intermediate values used in the function will be inferred by Haskell, and you
+don't have to state their types explicitly.
+
+Type inference is a killer feature of Haskell, and once you get used to it, you
+won't want to work without it.
+
+## Type Declarations
+
+In Haskell, you declare types on a separate line from where they are used. 
+Even though I said above that you usually only declare types for functions, in
+these examples we are going to declare them for single values.
+
+Type names are always capitalized in Haskell, and value (or function) names 
+are always lower-case.
+
+### Syntax
+
+```
+valueName :: Type
+valueName = definition
+```
+
+## Primitive types
+
+### Boolean
+
+Haskell has booleans, and they can either be `True` or `False`
+
+```haskell
+imTrue :: Bool
+imTrue = True
+
+imFalse :: Bool
+imFalse = False
+```
+
+### Integer
+
+Haskell has two common integer types, `Int` and `Integer`
+
+#### `Int`
+
+A fixed-precision type, with at least the range `[-2^29 .. 2^29-1]`, or 30 bits
+It is commonly implemented as 32 or 64 bits
+
+```haskell
+positiveInt :: Int
+positiveInt = 42
+
+negativeInt :: Int
+negativeInt = -37
+```
+
+The fact that `Int`'s size is limited means it is subject to overflows. These 
+occur without warning, so if they are a concern, use `Integer`
+
+#### `Integer`
+
+An arbitrary-precision type. Its size is only limited by the memory on your
+machine.
+
+```haskell
+positiveInteger :: Integer
+positiveInteger = 42
+
+negativeInteger :: Integer
+negativeInteger = -37
+```
+
```

Comment:
Any chance you could add a little something about `Word`? I'm especially interested in promoting it as a type that can't be negative, instead of using `Int` and hoping for the best. It's also the occasion to evoke the *correct by construction* paradigm.

Reply:
👍

Reply:
Consider mentioning `Natural` instead of or together with `Word`?

Reply:
Yes!!

Reply:
And the fact that arithmetic underflows are handled differently:

When doing `0 - 1`

* Since `Word` is modular, it loops back to `18446744073709551615`
* In the case of `Natural`, an `arithmetic underflow` exception is thrown